hphp/runtime/vm/jit/vasm-lower.cpp

```diff
@@ -33,6 +33,55 @@ namespace HPHP { namespace jit {
 
 ///////////////////////////////////////////////////////////////////////////////
 
+namespace detail {
+
+jit::vector<int> computeDominatingConstraints(Vunit& unit) {
+  jit::vector<int> vregConstraints(unit.blocks.size(), -1);
+
+  // Build reverse postorder list of blocks
+  auto const rpo = sortBlocks(unit);
+
+  // Calculate each block's dominating vreg constraint. The lowerers will use
+  // this as the initial vreg constraint for each block, adjusting as they
+  // encounter vregallow/vregrestrict during lowering.
+  vregConstraints[rpo[0]] = 0;
+  for (auto b : rpo) {
+    // Iterate through the instructions updating the constraint accordingly.
+    auto con = vregConstraints[b];
+    for (size_t i = 0; i < unit.blocks[b].code.size(); ++i) {
+      auto& inst = unit.blocks[b].code[i];
+      if (inst.op == Vinstr::vregrestrict) {
+        con++;
+      } else if (inst.op == Vinstr::vregallow) {
+        con--;
+      }
+    }
+    assert_flog(con >= 0,
+		"Dominating constraint can't be unknown.");
```

Comment:
> If we're going to allow "nesting" of constraints, then we probably don't want this "unknown" behavior. In particular, if we're going to have vregallow{} and vregrestrict{} be intrinsic to PhysRegSaver, then we need to let the vregallow{} in the constructor change the level to -1, then have the vregrestrict{} change the level back to 0, so that PhysRegSaver will still work without being dominated by a vregrestrict{} (or more generally, without being in a context where the constraint level is >= 1).

Currently, `PhysRegSaver` is only used within a block and so this isn't an issue. This is because the scoped restriction level is only used when computing the dominating constraints. Once iterating over a block, then we rely on the simple bool, `allow_vreg`. I'm totally OK with allowing negative `vregConstraints` for the dominating constraints if you think it'll be an issue in the future. I'll follow up on IRC.

> Okay, so... why not change `allow_vreg` to be an int (`vreg_level` or such), and have `vregallow{}` and `vregrestrict{}` decrement and increment it as usual?

I thought of this, but I think it violates the semantic of `vregallow`. I think that nesting for the purpose of tracking to see if we're still in a `vregrestrict` scope is useful. I think that whenever we have a `vregallow`, that all of the included instructions should be allowed to use new vregs until the next `vregrestrict` closes that scope.

As an example, consider a case where the current dominating vreg constraint for a block is 3. If that blocks contains a sequence of `Vinstr`s that are enclosed by a `PhysRegSaver`, then they should be able to use new `vreg`s. But if we use the `vreg_level` approach while iterating over the `Vinstrs` in a block, then the level would only be 2 for that sequence.